From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Narayan Reddy <narayanr@nvidia.com>
Date: Mon, 21 Feb 2022 08:53:31 +0000
Subject: [PATCH 026/240] NVIDIA: SAUCE: net: phy: aqr113c: restore PHY
 configuration.

Restore PHY configuration once PHY receives
the magic packet

add ethernet wake source

Bug 3447132

Change-Id: I2497e570adb4a016c5152e13cf759d60c5e75f53
Signed-off-by: Narayan Reddy <narayanr@nvidia.com>
Signed-off-by: Revanth Kumar Uppala <ruppala@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/linux-stable/+/2906740

diff --git a/drivers/net/phy/aquantia/aquantia_main.c b/drivers/net/phy/aquantia/aquantia_main.c
index 41f3676c7f1e..6b7009e6d4ff 100644
--- a/drivers/net/phy/aquantia/aquantia_main.c
+++ b/drivers/net/phy/aquantia/aquantia_main.c
@@ -110,6 +110,8 @@
 #define MDIO_AN_RX_VEND_STAT3			0xe832
 #define MDIO_AN_RX_VEND_STAT3_AFR		BIT(0)
 
+#define MDIO_C22EXT_GBE_PHY_SGMII_TX_ALARM1		0xec20
+
 /* Sleep and timeout for checking if the Processor-Intensive
  * MDIO operation is finished
  */
@@ -290,8 +292,32 @@ static int aqr_config_intr(struct phy_device *phydev)
 
 static irqreturn_t aqr_handle_interrupt(struct phy_device *phydev)
 {
+	int reg, val, ret;
 	int irq_status;
 
+	reg = phy_read_mmd(phydev, MDIO_MMD_C22EXT, MDIO_C22EXT_GBE_PHY_SGMII_TX_ALARM1);
+	if ((reg & MDIO_C22EXT_SGMII0_MAGIC_PKT_FRAME_MASK) ==
+	    MDIO_C22EXT_SGMII0_MAGIC_PKT_FRAME_MASK) {
+		/* Disable the WoL enable bit */
+		val = phy_read_mmd(phydev, MDIO_MMD_AN, MDIO_AN_RESVD_VEND_PROV);
+		val &= ~MDIO_MMD_AN_WOL_ENABLE;
+		ret = phy_write_mmd(phydev, MDIO_MMD_AN, MDIO_AN_RESVD_VEND_PROV, val);
+		if (ret < 0)
+			return IRQ_NONE;
+
+		/* Restore the SERDES/System Interface back to the XFI mode */
+		ret = phy_write_mmd(phydev, MDIO_MMD_VEND1,
+				    VEND1_GLOBAL_SYS_CONFIG_100M, VEND1_GLOBAL_SYS_CONFIG_XFI);
+		if (ret < 0)
+			return IRQ_NONE;
+		ret = phy_write_mmd(phydev, MDIO_MMD_VEND1,
+				    VEND1_GLOBAL_SYS_CONFIG_1G, VEND1_GLOBAL_SYS_CONFIG_XFI);
+		if (ret < 0)
+			return IRQ_NONE;
+		/* restart auto-negotiation */
+		return genphy_c45_restart_aneg(phydev);
+	}
+
 	irq_status = phy_read_mmd(phydev, MDIO_MMD_AN,
 				  MDIO_AN_TX_VEND_INT_STATUS2);
 	if (irq_status < 0) {
-- 
2.55.0

